### ISSUE TITLE
Expect reads and returns data out of order

### ISSUE THREAD

**Issue by viralmutant** [sshnet/SSH.NET#444]
I have a test script which write a command to host and then immediately reads it to ensure that command was executed
```
// Writing data
SshShellStream.Write(command + "\n");
SshShellStream.Flush();
// Reading what was written
bool rc = SshShellStream.Expect(regex, timeout == null ? TimeSpan.FromMinutes(30) : (TimeSpan)timeout, out result);
if (!rc)
{
string msg = "Ssh shell Expect time out! " + result;
LogEvents.OnDebug(msg);
throw new Exception(msg);
}
```

The code throws exception occasionally and the result is printed as
`System.Exception: Ssh shell Expect time out! 
sharsysadmin@my-local-host# e dest`

The output expected was
`sysadmin@my-local-host# share destroy`

Which is same as what was returned, but was out of order

**Comment by keithallenjackson** [comment]
First, SSH standard does not use `\n` for sending an <kbd>Enter</kbd>, but uses `\r`. Because of this, the remote machine is probably not getting the simulated <kbd>Enter</kbd> key press and the command never runs.

**Comment by viralmutant** [comment]
The command is executed on the remote machine. Just the read of the executed command is intermittently out of order. Could this be network latency issue ?

**Comment by keithallenjackson** [comment]
No, SSH uses TCP which guarantees delivery (through acknowledgements and retransmissions) and order as long as the connection is up.

Make sure you wait for the prompt before sending your command by using Expect and a regex.

**Comment by keithallenjackson** [comment]
```csharp

var promptRegex = new Regex(@"(?sm)(?<banner>.*)\s+(?<prompt>.+?[>#])\s+\k<prompt>");

Shell.Write("\r");

var output = Shell.Expect(promptRegex, TimeSpan.FromSeconds(30));
var match = promptRegex.Match(output);
var prompt = match.Groups["prompt"].Value;
var banner = match Groups["banner"].Value;

Shell.WriteLine("some command");
var commandOutput = Shell.Expect(prompt, TimeSpan.FromSeconds(30));

```

**Comment by viralmutant** [comment]
Sending the command is not causing the issue in my case.
I can see on the server end that command was received.

It's the .Expect() TIMEOUT which reads what was sent.

The shell in my case is different and requires `\n` instead of `\r`
This is what a comment in code reads:
`// the library comes with writeline method using "\r\n" which will not work with our shell`

**Comment by viralmutant** [comment]
Thanks I'll check. It seems when the error occurs, during a long running routine, the server is unable to receive the command.

After an Expect() matches and returns the string, how do I clear the rest of the stream ?
Also, after Write() do I need Flush() ?

I somehow feel, my issue is one of the already reported bugs :(
https://github.com/sshnet/SSH.NET/issues/180
https://github.com/sshnet/SSH.NET/issues/301
https://github.com/sshnet/SSH.NET/issues/302

What does `\s+\k<prompt>` towards the end of the regex you suggested means ?

**Comment by keithallenjackson** [comment]
\s+  means 1 or more whitespace characters
\k<prompt>  is a back reference to the named capture group "prompt"

**Comment by viralmutant** [comment]
Is there a workaround for any of those mentioned bugs ?

**Comment by WojciechNagorski** [comment]
This issue should be fixed by #1322
